src/servers/src/grpc/utils.rs

```diff
@@ -0,0 +1,156 @@
+pub fn extract_header<'a>(headers: &'a MetadataMap, keys: &[&str]) -> TonicResult<Option<&'a str>> {
+    let mut value = None;
+    for key in keys {
+        if let Some(v) = headers.get(*key) {
+            value = Some(v);
+            break;
+        }
+    }
+
+    let Some(v) = value else {
+        return Ok(None);
+    };
+    let Ok(v) = std::str::from_utf8(v.as_bytes()) else {
+        return Err(InvalidParameterSnafu {
+            reason: "expect valid UTF-8 value",
+        }
+        .build()
+        .into());
+    };
+    Ok(Some(v))
+}
+
+/// Helper function to extract the header from the metadata and authenticate the user.
+pub async fn check_auth(
+    user_provider: Option<UserProviderRef>,
+    headers: &MetadataMap,
+    query_ctx: QueryContextRef,
+) -> TonicResult<bool> {
+    if user_provider.is_none() {
+        return Ok(true);
+    }
+
+    let auth_schema = extract_header(
+        headers,
+        &[AUTHORIZATION_HEADER, http::header::AUTHORIZATION.as_str()],
+    )?
+    .map(|x| x.try_into())
+    .transpose()?
+    .map(|x: crate::http::authorize::AuthScheme| x.into());
+
+    let auth_schema = auth_schema.context(NotFoundAuthHeaderSnafu)?;
+    let header = RequestHeader {
+        authorization: Some(AuthHeader {
+            auth_scheme: Some(auth_schema),
+        }),
+        catalog: query_ctx.current_catalog().to_string(),
+        schema: query_ctx.current_schema(),
+        ..Default::default()
+    };
+
+    match auth(user_provider, Some(&header), &query_ctx).await {
+        Ok(user_info) => {
+            query_ctx.set_current_user(user_info);
+            Ok(true)
+        }
+        Err(_) => Err(Status::unauthenticated("auth failed")),
+    }
+}
+
+/// Authenticate the user based on the header and query context.
+pub async fn auth(
+    user_provider: Option<UserProviderRef>,
+    header: Option<&RequestHeader>,
+    query_ctx: &QueryContextRef,
+) -> Result<UserInfoRef> {
+    let Some(user_provider) = user_provider else {
+        return Ok(auth::userinfo_by_name(None));
+    };
+
+    let auth_scheme = header
+        .and_then(|header| {
+            header
+                .authorization
+                .as_ref()
+                .and_then(|x| x.auth_scheme.clone())
+        })
+        .context(NotFoundAuthHeaderSnafu)?;
+
+    match auth_scheme {
+        AuthScheme::Basic(api::v1::Basic { username, password }) => user_provider
+            .auth(
+                Identity::UserId(&username, None),
+                Password::PlainText(password.into()),
+                query_ctx.current_catalog(),
+                &query_ctx.current_schema(),
+            )
+            .await
+            .context(AuthSnafu),
+        AuthScheme::Token(_) => Err(UnsupportedAuthScheme {
+            name: "Token AuthScheme".to_string(),
+        }),
+    }
+    .inspect_err(|e| {
+        METRIC_AUTH_FAILURE
+            .with_label_values(&[e.status_code().as_ref()])
+            .inc();
+    })
+}
```

Comment:
it's better to rename this module, auth is not a "utility"

Reply:
Renamed to `context_auth`